src/core/geometry/qgspolyhedralsurface.h

```diff
@@ -0,0 +1,341 @@
+    int vertexNumberFromVertexId( QgsVertexId id ) const override;
+    bool isEmpty() const override SIP_HOLDGIL;
+    double closestSegment( const QgsPoint &pt, QgsPoint &segmentPt SIP_OUT, QgsVertexId &vertexAfter SIP_OUT, int *leftOf SIP_OUT = nullptr, double epsilon = 4 * std::numeric_limits<double>::epsilon() ) const override;
+
+    bool nextVertex( QgsVertexId &id, QgsPoint &vertex SIP_OUT ) const override;
+    void adjacentVertices( QgsVertexId vertex, QgsVertexId &previousVertex SIP_OUT, QgsVertexId &nextVertex SIP_OUT ) const override;
+    bool hasCurvedSegments() const final;
+
+    /**
+     * Returns a geometry without curves. Caller takes ownership
+     * \param tolerance segmentation tolerance
+     * \param toleranceType maximum segmentation angle or maximum difference between approximation and curve
+    */
+    QgsAbstractGeometry *segmentize( double tolerance = M_PI_2 / 90, SegmentationToleranceType toleranceType = MaximumAngle ) const override SIP_FACTORY;
+
+    /**
+     * Returns approximate rotation angle for a vertex. Usually average angle between adjacent segments.
+     * \param vertex the vertex id
+     * \returns rotation in radians, clockwise from north
+     */
+    double vertexAngle( QgsVertexId vertex ) const override;
+
+    int vertexCount( int part = 0, int ring = 0 ) const override;
+    int ringCount( int part = 0 ) const override SIP_HOLDGIL;
+    int partCount() const override SIP_HOLDGIL;
+    QgsPoint vertexAt( QgsVertexId id ) const override;
+    double segmentLength( QgsVertexId startVertex ) const override;
+
+    bool addZValue( double zValue = 0 ) override;
+    bool addMValue( double mValue = 0 ) override;
+    bool dropZValue() override;
+    bool dropMValue() override;
+    void swapXy() override;
+
+    QgsMultiSurface *toCurveType() const override SIP_FACTORY;
+
+    bool transform( QgsAbstractGeometryTransformer *transformer, QgsFeedback *feedback = nullptr ) override;
+
+    //! Converts a polyhedral surface to a multipolygon.
+    QgsMultiPolygon *toMultiPolygon() const SIP_FACTORY;
+
+#ifndef SIP_RUN
+    void filterVertices( const std::function< bool( const QgsPoint & ) > &filter ) override;
+    void transformVertices( const std::function< QgsPoint( const QgsPoint & ) > &transform ) override;
+
+    /**
+     * Cast the \a geom to a QgsPolyhedralSurface.
+     * Should be used by qgsgeometry_cast<QgsPolyhedralSurface *>( geometry ).
+     *
+     * \note Not available in Python. Objects will be automatically be converted to the appropriate target type.
+     */
+    inline static const QgsPolyhedralSurface *cast( const QgsAbstractGeometry *geom )
+    {
+      if ( !geom )
+        return nullptr;
+
+      const Qgis::WkbType flatType = QgsWkbTypes::flatType( geom->wkbType() );
+      if ( flatType == Qgis::WkbType::PolyhedralSurface
+           || flatType == Qgis::WkbType::TIN
+           || flatType == Qgis::WkbType::Polygon
+           || flatType == Qgis::WkbType::Triangle )
```

Comment:
you can't cast a Polygon* or Triangle* to a PolyhedralSurface*
```suggestion
           || flatType == Qgis::WkbType::TIN )
```